File: api/src/main/java/com/datastrato/gravitino/file/FilesetDataOperationCtx.java
/*
 * Copyright 2024 Datastrato Pvt Ltd.
 * This software is licensed under the Apache License version 2.
 */
package com.datastrato.gravitino.file;

import com.datastrato.gravitino.annotation.Evolving;
import java.util.Map;

/**
 * An interface representing a fileset data operation context. This interface defines some
 * information need to report to the server.
 *
 * <p>{@link FilesetDataOperationCtx} defines the basic properties of a fileset data operation
 * context object.
 */
@Evolving
public interface FilesetDataOperationCtx {
  /**
   * @return The sub path which is operated by the data operation. For example: /test.txt, test.txt.
   */
  String subPath();

  /** @return The data operation type. For example: CREATE, APPEND, DELETE, etc. */
  FilesetDataOperation operation();

  /**
   * @return The client type of the data operation happened. For example: HADOOP_GVFS, PYTHON_GVFS,
   *     UNKNOWN.
   */
  ClientType clientType();

  /** @return The client ip of the data operation. */
  String ip();

  /**
   * @return The source engine type of the data operation happened. For example: SPARK, FLINK,
   *     CLOUDML, UNKNOWN.
   */
  SourceEngineType sourceEngineType();

  /** @return The application id of the data operation. */
  String appId();

  /** @return The extra info of the data operation. */
  Map<String, String> extraInfo();
}

File: clients/client-python/gravitino/api/fileset_data_operation_ctx.py
"""
Copyright 2024 Datastrato Pvt Ltd.
This software is licensed under the Apache License version 2.
"""

from abc import ABC, abstractmethod

from typing import Dict

from gravitino.api.client_type import ClientType
from gravitino.api.fileset_data_operation import FilesetDataOperation
from gravitino.api.source_engine_type import SourceEngineType


class FilesetDataOperationCtx(ABC):
    """An interface representing a fileset data operation context. This interface defines some
    information need to report to the server.
    """

    @abstractmethod
    def sub_path(self) -> str:
        """The sub path which is operated by the data operation.

        Returns:
            the sub path which is operated by the data operation.
        """
        pass

    @abstractmethod
    def operation(self) -> FilesetDataOperation:
        """The data operation type.

        Returns:
            the data operation type.
        """
        pass

    @abstractmethod
    def client_type(self) -> ClientType:
        """The client type of the data operation.

        Returns:
            the client type of the data operation.
        """
        pass

    @abstractmethod
    def ip(self) -> str:
        """The client ip of the data operation.

        Returns:
            the client ip of the data operation.
        """
        pass

    @abstractmethod
    def source_engine_type(self) -> SourceEngineType:
        """The source type of the data operation.

        Returns:
            the source type of the data operation.
        """
        pass

    @abstractmethod
    def app_id(self) -> str:
        """The application id of the data operation.

        Returns:
            the application id of the data operation.
        """
        pass

    @abstractmethod
    def extra_info(self) -> Dict[str, str]:
        """The extra info of the data operation.

        Returns:
            the extra info of the data operation.
        """
        pass
